xtext.cpp

```diff
@@ -0,0 +1,2157 @@
+                        // What we wrote is a high surrogate, and there's a possible low surrogate following
+                        if ( HEAD_CHECK(0xF0, 0xE0) && IS_FOLLOWING(1) && IS_FOLLOWING(2) ) { // is a valid 3-bytes sequence
+                            uint32_t next = HEAD_BYTE(0x0F, 12) | CONT_BYTE(1,6) | CONT_BYTE(2,0);
+                            if (next >= 0xDC00 && next <= 0xDFFF) { // is a low surrogate: valid surrogates sequence
+                                // Override what we wrote with the codepoint for this high+low surrogates sequence
+                                *p = 0x10000 + ((*p & 0x3FF)<<10) + (next & 0x3FF);
+                                s += 3;
+                            }
+                        }
+                    }
+                    // todo: deal with invalide surrotage sequences
+                }
+            }
+        }
+        else if ( HEAD_CHECK(0xF8, 0xF0) ) {
+            if ( HAS_FOLLOWUP(3) && IS_FOLLOWING(1) && IS_FOLLOWING(2) && IS_FOLLOWING(3) ) {
+                if ( do_decode )
+                    *p = HEAD_BYTE(0x07, 18) | CONT_BYTE(1,12) | CONT_BYTE(2,6) | CONT_BYTE(3,0);
+                s += 4;
+                valid = true;
+            }
+        }
+        // else: invalid first byte in UTF-8 sequence
+
+        if ( !valid ) {
+            if (do_decode)
+                *p = REPLACEMENT_CHAR;
+            s++;
+            is_valid = false;
+        }
+        if ( do_decode ) {
+	    // Try to detect if we have RTL chars, so that if we don't have any,
+	    // we don't need to invoke expensive fribidi processing.
+	    if ( !has_rtl )
+                has_rtl = is_unicodepoint_rtl(*p);
+            p++;
+        }
+        ucount++;
+    }
+    return ucount;
+}
+
+// ==============================================
+// Flags, data structures, and global variables
+
+#define HINT_DIRECTION_IS_RTL   0x0001 /// segment direction is RTL
+#define HINT_BEGINS_PARAGRAPH   0x0002 /// segment is at start of paragraph
+#define HINT_ENDS_PARAGRAPH     0x0004 /// segment is at end of paragraph
+
+#define CHAR_CAN_WRAP_AFTER        0x0001
+#define CHAR_MUST_BREAK_AFTER      0x0002
+#define CHAR_SKIP_ON_BREAK         0x0004
+#define CHAR_CAN_EXTEND_WIDTH      0x0008
+#define CHAR_IS_CLUSTER_TAIL       0x0010
+#define CHAR_IS_RTL                0x0020
+#define CHAR_SCRIPT_CHANGE         0x0040
+#define CHAR_IS_PARA_START         0x0100
+#define CHAR_IS_PARA_END           0x0200
+#define CHAR_PARA_IS_RTL           0x0400 /// to know the line with this char is part
+                                           /// of a paragraph with main dir RTL
+
+// Info, after measure(), about each m_text char
+typedef struct {
+    unsigned short flags;
+    signed short   width;
+} xtext_charinfo_t;
+
+// Glyph info when shaping a line (to be returned to Lua as a table of tables)
+// (16 bytes, making our static s_shape_result[MAX_LINE_GLYPHS] a 128Kb buffer)
+typedef struct {
```

Comment:
Random comment about packing rules if you want to micro-optimize memory usage.

Ordering the struct by data type's width (i.e., wider to narrower, or vice-versa) usually works well enough.

May hurt readability (and/or actual performance, depending on what's used, how and when).

C is fun :D.

Reply:
And I just noticed the bitfields, which will by definition cause padding anyway ^^.

Reply:
Speaking of the bitfields, any specific reason for the weird 1 + 1 + explicit padding instead of using two bools?

Reply:
Limiting the size of the whole (a bit is 1 bit, a bool is 8 bits) - and trying to have the big ones aligned. I rewrote that one a few times, each time printing its sizeof :)

```
__________0
    int text_index;    // original index in m_text
__________4
    uint16_t glyph;    // glyph index in font
    unsigned char font_num;
    unsigned is_rtl:1;
    unsigned can_extend:1;
    unsigned _unused:6;
__________8
    signed short  x_advance;  // aligned
    signed short  x_offset;
__________12
    signed short  y_offset; // aligned
    unsigned char cluster_len;
    unsigned is_cluster_start:1; // could have been moved with the others
                   // but I like it near cluster_len, and I have the room
__________16
```

Reply:
Err, `sizeof(bool) == 1`?

Reply:
Argh, that bool is 8 bits thingy appear to be a C++ shenanigan. >_<".

Reply:
Yes. It's not `0.125` :)
```c
typedef struct {
    int text_index;    // original index in m_text
    uint16_t glyph;    // glyph index in font
    signed short  x_advance;
    signed short  x_offset;
    signed short  y_offset;
    unsigned char font_num;
    unsigned char cluster_len;
    bool is_cluster_start;
    bool is_rtl;
    bool can_extend; // one byte too much, these last 3 are NOT merged into one byte
} xtext_shapeinfo_t;
```
ùakes its sizeof() = 20.
My previous with :bits kept it 16.

Reply:
Oh, yeah, duh, the bitfields essentially allows you to fit two bools in one byte.

The larger inconsistency in your last example is probably because of enforced alignment to the wider type (int == 4).